cpp/src/gandiva/decimal_format_holder.cc

```diff
@@ -0,0 +1,109 @@
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "gandiva/decimal_format_holder.h"
+
+namespace gandiva {
+static bool IsArrowStringLiteral(arrow::Type::type type) {
+  return type == arrow::Type::STRING || type == arrow::Type::BINARY;
+}
+inline double get_scale_multiplier(int32_t scale) {
+  static const double values[] = {1.0,
+                                  10.0,
+                                  100.0,
+                                  1000.0,
+                                  10000.0,
+                                  100000.0,
+                                  1000000.0,
+                                  10000000.0,
+                                  100000000.0,
+                                  1000000000.0,
+                                  10000000000.0,
+                                  100000000000.0,
+                                  1000000000000.0,
+                                  10000000000000.0,
+                                  100000000000000.0,
+                                  1000000000000000.0,
+                                  10000000000000000.0,
+                                  100000000000000000.0,
+                                  1000000000000000000.0,
+                                  10000000000000000000.0};
+  if (scale >= 0 && scale < 20) {
+    return values[scale];
+  }
+  return static_cast<double>(powl(10.0, scale));
+}
+
+inline double round_decimal_digits(double to_round, int32_t scale) {
+  double scale_multiplier = get_scale_multiplier(scale);
+  return trunc(to_round * scale_multiplier + ((to_round >= 0) ? 0.5 : -0.5)) /
+         scale_multiplier;
+}
+
+Status DecimalFormatHolder::Make(const FunctionNode& node,
+                                 std::shared_ptr<DecimalFormatHolder>* holder) {
+  ARROW_RETURN_IF(node.children().size() != 2,
+                  Status::Invalid("'to_number' function requires two parameters"));
+  auto literal = dynamic_cast<LiteralNode*>(node.children().at(1).get());
+
+  ARROW_RETURN_IF(literal == nullptr,
+                  Status::Invalid("'to_number' function requires a"
+                                  " literal as the second parameter"));
+
+  auto literal_type = literal->return_type()->id();
+  ARROW_RETURN_IF(
+      !IsArrowStringLiteral(literal_type),
+      Status::Invalid(
+          "'to_number' function requires a string literal as the second parameter"));
+  return DecimalFormatHolder::Make(arrow::util::get<std::string>(literal->holder()),
+                                   holder);
+}
+
+Status DecimalFormatHolder::Make(const std::string& decimal_format,
+                                 std::shared_ptr<DecimalFormatHolder>* holder) {
+  auto lholder = std::shared_ptr<DecimalFormatHolder>(
+      new DecimalFormatHolder(decimal_format.c_str(), decimal_format.size()));
```

Comment:
isn't this a bit dangerous to have an object have a pointer to a string c_str. What if string decimal_format was constructed in the stack by the caller of Make?
Is it always guaranteed that the decimal_format object passed here has a lifetime more than the DecimalFormatHolder object?
As far as I can see the shared_ptr is against the DecimalFormatHolder and not against decimal_format. So if decimal_format goes out of scope. decimal_format.c_str() will be pointing to a freed memory location.